Avoid the propagation of RequiresUnreferencedCode to nested classes

tlakollo opened the pull request:
Avoid the propagation of RequiresUnreferencedCode to nested classes
When suppressing warnings only ask the declaring type instead of the effective type
Fixes #2151

vitek-karas commented on src/linker/Linker.Steps/MarkStep.cs:
```diff
@@ -1768,9 +1768,13 @@ protected internal virtual TypeDefinition MarkType (TypeReference reference, Dep
-				!_context.Annotations.TryGetEffectiveRequiresUnreferencedCodeAttributeOnType (type, out RequiresUnreferencedCodeAttribute _) &&
-				_context.Annotations.TryGetEffectiveRequiresUnreferencedCodeAttributeOnType (_context.TryResolve (type.BaseType), out RequiresUnreferencedCodeAttribute effectiveRequiresUnreferencedCode)) {
+			if (((type.DeclaringType is not null &&
+				!_context.Annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (type.DeclaringType)) ||
```
Just to capture our offline discussion:
```C#
[RequiresUnreferencedCode("test")]
class RUCBase {}

[RequiresUnreferencedCode("test")]
class OuterType
{
    // Expecting warning here - as the RUC on OuterType should not apply to anything on nested types
    class Derived : RUCBase
    {
    }
}
```

agocke reviewed (approved):
LGTM

tlakollo commented:
Changed the different checks to call DoesMethodRequireUnreferencedCode to prevent having inconsistent calls, also moved the DoesMethodRequireUnreferencedCode method from MarkStep to the Annotations file

mateoatr commented on src/linker/Linker.Steps/MarkStep.cs:
```diff
@@ -2903,13 +2903,13 @@ internal bool ShouldSuppressAnalysisWarningsForRequiresUnreferencedCode ()
 			IMemberDefinition originMember = currentOrigin.MemberDefinition;
-			if (suppressionContextMember != originMember && originMember != null &&
-				Annotations.DoesMethodRequireUnreferencedCode (originMember, out _))
+			if (originMember is not null && originMember is MethodDefinition && suppressionContextMember != originMember &&
```
It is sufficient to have `originMember is MethodDefinition` (same above for suppressionContextMember.
